Modify Videopipeline to send H264 buffers instead of RTP packets

raveendra-karu opened the pull request:
#### Summary

Video pipeline in the camera source sending RTP H264 packets to the consumers (ex: WebRTC Transport).
To Support PushAV (Clip record), PushAVTransport needs H264 frame buffers instead of RTP H264 packets.
Modified Video pipeline to send H264 frame buffers instead of RTP packets which works for both WebRTC and PushAV Transports
Also, updated webrtc video track to packetize the frame buffers and send over the libdatachannel connection.

#### Testing
* Build & Start Camera Application
```./scripts/examples/gn_build_example.sh examples/camera-app/linux out/debug```
```rm -rf /tmp/chip_*; ./out/debug/chip-camera-app```

* Build & Start Camera controller
```./scripts/build/build_examples.py --target linux-x64-camera-controller build```
```./out/linux-x64-camera-controller/chip-camera-controller```

* Run WebRTC live commands
```liveview start 1```
* Observe that the live stream is playing in the controller window

raveendra-karu committed:
Add RTP Packetizer for sending live video buffers.

Author: Dhruvkumar <kakadiya.b@samsung.com>

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Modify Video pipeline to send  H264 video buffers without RTP packetization.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Refactor and clean up.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fix Typo.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

yufengwangca commented:
The liveview function was broken with this PR, and I got the following errors

```
[1755213646.153] [355242:355242] [DMG] AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_0551 e=1 p=o r=i
[1755213646.153] [355242:355242] [DMG] AccessControl: allowed
[1755213646.153] [355242:355242] [DMG] AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_0551 e=1 p=m r=i
[1755213646.153] [355242:355242] [DMG] AccessControl: allowed
[1755213646.153] [355242:355242] [DMG] Received command for Endpoint=1 Cluster=0x0000_0551 Command=0x0000_0003
[1755213646.153] [355242:355242] [ZCL] CameraAVStreamMgmt[ep=1]: Allocating Video Stream

(chip-camera-app:355242): GLib-GObject-CRITICAL **: 16:20:46.166: value "((GstX264EncTune) Still image | Fast decode | 3590710520)" of type 'GstX264EncTune' is invalid or out of range for property 'tune' of type 'GstX264EncTune'

(chip-camera-app:355242): GLib-GObject-CRITICAL **: 16:20:46.166: g_object_set_is_valid_property: object class 'GstX264Enc' has no property named 'profile'
[1755213646.170] [355242:355242] [CAM] Starting video stream (id=1): 640×360 @ 30fps
[1755213646.170] [355242:355242] [CAM] Requesting PLAYING …
[1755213651.170] [355242:355242] [CAM] Video pipeline did not reach PLAYING state.
[1755213651.263] [355242:355242] [CAM] Setting per stream viewport for stream 1.
[1755213651.263] [355242:355242] [CAM] New viewport. x1=0, x2=1920, y1=0, y2=1080.
[1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0552 update version to b5b87b
[1755213651.263] [355242:355242] [CAM] Attribute changed for AttributeId = 0x0000_000C
[1755213651.263] [355242:355242] [CAM] Unknown Attribute with AttributeId = 0x0000_000C
[1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0551 update version to 146077bf
[1755213651.263] [355242:355242] [CAM] Attribute changed for AttributeId = 0x0000_000F
[1755213651.263] [355242:355242] [CAM] Unknown Attribute with AttributeId = 0x0000_000F
[1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0551 update version to 146077c0
[1755213651.263] [355242:355242] [DMG] Command handler moving to [NewRespons]
[1755213651.263] [355242:355242] [DMG] Command handler moving to [ Preparing]

```

yufengwangca reviewed (changes requested):
[1755213646.153] [355242:355242] [DMG] AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_0551 e=1 p=o r=i
[1755213646.153] [355242:355242] [DMG] AccessControl: allowed
[1755213646.153] [355242:355242] [DMG] AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_0551 e=1 p=m r=i
[1755213646.153] [355242:355242] [DMG] AccessControl: allowed
[1755213646.153] [355242:355242] [DMG] Received command for Endpoint=1 Cluster=0x0000_0551 Command=0x0000_0003
[1755213646.153] [355242:355242] [ZCL] CameraAVStreamMgmt[ep=1]: Allocating Video Stream

(chip-camera-app:355242): GLib-GObject-CRITICAL **: 16:20:46.166: value "((GstX264EncTune) Still image | Fast decode | 3590710520)" of type 'GstX264EncTune' is invalid or out of range for property 'tune' of type 'GstX264EncTune'

(chip-camera-app:355242): GLib-GObject-CRITICAL **: 16:20:46.166: g_object_set_is_valid_property: object class 'GstX264Enc' has no property named 'profile'
[1755213646.170] [355242:355242] [CAM] Starting video stream (id=1): 640×360 @ 30fps
[1755213646.170] [355242:355242] [CAM] Requesting PLAYING …
[1755213651.170] [355242:355242] [CAM] Video pipeline did not reach PLAYING state.
[1755213651.263] [355242:355242] [CAM] Setting per stream viewport for stream 1.
[1755213651.263] [355242:355242] [CAM] New viewport. x1=0, x2=1920, y1=0, y2=1080.
[1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0552 update version to b5b87b
[1755213651.263] [355242:355242] [CAM] Attribute changed for AttributeId = 0x0000_000C
[1755213651.263] [355242:355242] [CAM] Unknown Attribute with AttributeId = 0x0000_000C
[1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0551 update version to 146077bf
[1755213651.263] [355242:355242] [CAM] Attribute changed for AttributeId = 0x0000_000F
[1755213651.263] [355242:355242] [CAM] Unknown Attribute with AttributeId = 0x0000_000F
[1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0551 update version to 146077c0
[1755213651.263] [355242:355242] [DMG] Command handler moving to [NewRespons]
[1755213651.263] [355242:355242] [DMG] Command handler moving to [ Preparing]

raveendra-karu committed:
Rebased to latest code.

Increase timeout to get pipeline state.
Co-Author: Dhruvkumar <kakadiya.b@samsung.com>

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fix pipeline errors.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

pidarped commented on examples/camera-app/linux/src/camera-device.cpp:
```diff
@@ -678,7 +672,7 @@ CameraError CameraDevice::StartVideoStream(uint16_t streamID)
     // Wait for the pipeline to reach the PLAYING state
     GstState state;
-    gst_element_get_state(videoPipeline, &state, nullptr, 5 * GST_SECOND);
+    gst_element_get_state(videoPipeline, &state, nullptr, 6 * GST_SECOND);
```
What is this choice of 6 seconds based on? Better to move to a #define at the top with a comment explaining the choice.

raveendra-karu replied:
Done.  Retained 5 secs itself as 6 secs not much difference.

pidarped commented on examples/camera-app/linux/src/webrtc-libdatachannel.cpp:
```diff
@@ -25,6 +25,7 @@ namespace {
 // Constants
 constexpr int kVideoH264PayloadType = 96;
 constexpr int kVideoBitRate         = 3000;
+constexpr int kSSRC                 = 42;
```
The SSRC should be unique for a stream source, right? Should this be randomly chosen?

raveendra-karu replied:
Yes, you are correct. Unique ssrc is preferred.
As only 1 video track used per peerconnection is sued, having fixed ssrc should be okay in this scenario. Keeping it as is for now.

raveendra-karu committed:
Set Max fragment size to 1200.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by clang-format

raveendra-karu commented on examples/camera-app/linux/src/webrtc-libdatachannel.cpp:
```diff
@@ -104,13 +105,37 @@ const char * GetGatheringStateStr(rtc::PeerConnection::GatheringState state)
-    LibDataChannelTrack(std::shared_ptr<rtc::Track> track) : mTrack(track) {}
+    LibDataChannelTrack(std::shared_ptr<rtc::Track> track) : mTrack(track) { InitH264Packetizer(); }
+    // Initialize libdatachannel's RTP packetizer for H.264
+    void InitH264Packetizer()
```
Ideally yes, As I verified packet size received at controller side on Linux machine with ethernet connection, MTU is going up to 1400.  Set 1200 as maxFragmentsize (including RTP headers) and updated the PR

yufengwangca commented:
> > [1755213646.153] [355242:355242] [DMG] AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_0551 e=1 p=o r=i [1755213646.153] [355242:355242] [DMG] AccessControl: allowed [1755213646.153] [355242:355242] [DMG] AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_0551 e=1 p=m r=i [1755213646.153] [355242:355242] [DMG] AccessControl: allowed [1755213646.153] [355242:355242] [DMG] Received command for Endpoint=1 Cluster=0x0000_0551 Command=0x0000_0003 [1755213646.153] [355242:355242] [ZCL] CameraAVStreamMgmt[ep=1]: Allocating Video Stream
> > (chip-camera-app:355242): GLib-GObject-CRITICAL **: 16:20:46.166: value "((GstX264EncTune) Still image | Fast decode | 3590710520)" of type 'GstX264EncTune' is invalid or out of range for property 'tune' of type 'GstX264EncTune'
> > (chip-camera-app:355242): GLib-GObject-CRITICAL **: 16:20:46.166: g_object_set_is_valid_property: object class 'GstX264Enc' has no property named 'profile' [1755213646.170] [355242:355242] [CAM] Starting video stream (id=1): 640×360 @ 30fps [1755213646.170] [355242:355242] [CAM] Requesting PLAYING … [1755213651.170] [355242:355242] [CAM] Video pipeline did not reach PLAYING state. [1755213651.263] [355242:355242] [CAM] Setting per stream viewport for stream 1. [1755213651.263] [355242:355242] [CAM] New viewport. x1=0, x2=1920, y1=0, y2=1080. [1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0552 update version to b5b87b [1755213651.263] [355242:355242] [CAM] Attribute changed for AttributeId = 0x0000_000C [1755213651.263] [355242:355242] [CAM] Unknown Attribute with AttributeId = 0x0000_000C [1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0551 update version to 146077bf [1755213651.263] [355242:355242] [CAM] Attribute changed for AttributeId = 0x0000_000F [1755213651.263] [355242:355242] [CAM] Unknown Attribute with AttributeId = 0x0000_000F [1755213651.263] [355242:355242] [DMG] Endpoint 1, Cluster 0x0000_0551 update version to 146077c0 [1755213651.263] [355242:355242] [DMG] Command handler moving to [NewRespons] [1755213651.263] [355242:355242] [DMG] Command handler moving to [ Preparing]
> 
> Hi @yufengwangca Checked the pipeline with testsrc and h/w src, sometimes it is taking time to start pipeline. Hence increased timeout. Could you please check and review

Just test the latest update, I still see the old error when I start the live view.

```
[1756415733.565] [168476:168476] [CAM] Starting new video stream with ID: 1

(chip-camera-app:168476): GLib-GObject-CRITICAL **: 14:15:33.575: value "((GstX264EncTune) Still image | Fast decode | 339522800)" of type 'GstX264EncTune' is invalid or out of range for property 'tune' of type 'GstX264EncTune'
[1756415733.577] [168476:168476] [CAM] Starting video stream (id=1): 640×360 @ 30fps
[1756415733.577] [168476:168476] [CAM] Requesting PLAYING …
[1756415738.578] [168476:168476] [CAM] Video pipeline did not reach PLAYING state.

```

Attached the complete log for your reference.
[PR_40594_live_view_08_28_25.txt](https://github.com/user-attachments/files/22033505/PR_40594_live_view_08_28_25.txt)

raveendra-karu committed:
Resolve GST video pipeline lauch warnings.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Implemented audio track and pipeline and enhanced video track

Author: Kakadiya Dhruvkumar Babubhai
Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Implemented audio track, pipeline and fix typo

Author: Kakadiya Dhruvkumar Babubhai
Signed-off-by: Raveendra Karu <r.karu@samsung.com>
